[Windows] Make XAudio2 work with Windows 8 and code improvements

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

Follow up PR of #25068 to make Xaudio output work on Windows 8.1 as well, and additional code improvements.

* The previous PR used the name returned by mmdevice enumeration to open an xaudio sink by name but those names are not accepted by Windows 8, which requires the names as enumerated on Xbox or UWP desktop build.

for example `\\?\SWD#MMDEVAPI#{0.0.0.00000000}.{a4ec9140-07bf-4031-9bb1-01f49002e272}#{e6327cad-dcec-4949-ae8a-991e976a79d2}`

Those names also work on Windows 10 and 11. Solved the compatibility issue with a refactor of the enumeration code, now shared between Xbox/UWP and desktop.

* Another issue was the different names and versions of the xaudio dll (2.8 or 2.9) depending on the OS. There is also a redistributable version that could have been installed by games on some computers (2.9 redist).
The minimum requirement for Kodi is version 2.8, bundled with Windows 8.1 so there is no need to install the redist package for a higher version. Instead, probe the system to dynamically load the highest version available (2.9 redist > 2,9 > 2.8)

* Finally, miscellaneous code improvements, one per commit.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Windows 8 is a supported platform that includes XAudio support but it did not work in Kodi.
Noticed various code issues.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Windows 8, default and named devices, builtin speakers / earbuds, hdmi audio devices.
Windows 10 and 11 desktop and UWP to check for no adverse effects, with speakers, hdmi, spdif, usb.

no bluetooth audio device available for testing.
Xbox not tested.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

Xaudio is more modern than DirectSound and is now available for all Kodi-supported Windows versions.
Still very new on desktop so not the default sink.

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

CrystalP committed:
[XAudio2] share Xbox audio device enumeration with desktop and Windows 8.1 compatibility

CrystalP committed:
[XAudio2] error handling improvements

CrystalP committed:
[XAudio2] skip testing of sampling rates not supported by XAudio2

CrystalP committed:
[XAudio2] avoid leak + fix voice creation for closest match

CrystalP commented:
Jenkins build this please

thexai reviewed (approved):
Code looks good.

Tested Windows x64 on Intel NUC with HDMI audio devices.
Tested on Xbox Series S.
Tested on desktop PC also with Bluetooth audio device.

No issues found.

CrystalP commented on xbmc/cores/AudioEngine/Sinks/AESinkXAudio.cpp:
```diff
@@ -573,19 +654,33 @@ bool CAESinkXAudio::InitializeInternal(std::string deviceId, AEAudioFormat &form
+        format.m_dataFormat = AE_FMT_U8;
+        break;
+
+      default:
+        format.m_dataFormat = AE_FMT_S16NE;
+        break;
```
@fritsch wonder if that's right - this is the same behavior as before for an unsupported number of bits, but doesn't seem right, would you agree?
Being in the default case means that the OS/audio device accepted the format but the sink doesn't fully support it.
I think that the sink should either return an error, or test the number of bits of the AE provided format earlier and when it's not a recognized number, go into the loop that uses the testFormats list to come up with a format supported by the sink. Latter option is almost as easy and seems more forward looking to me.

fritsch replied:
So in general during initializing we enumerate our supported formats in the sink and return those to AE. AE should then only call us with one of those after all.

The ordering is also important. Back the AE chose simply what was first in list.

On the next open though you can decide again, when overwriting the referenced format back to AE. Best is saving the supported list, updating on hot plug and using the biggest format available.

Pulse uses float, ALSA 32 bit int and android depending on multi channel or not 16 bit int or float.

Reason you can overwrite is also: sinks may not support all formats in all channel combinations. See android example.

CrystalP replied:
My testing and pretty much every log I come across show AE calling Initialize() with AE_FMT_FLOAT even when it was not added to the device during enumeration by the sink,
The WASAPI sink then always falls back to the predefined formats of testFormats to find and return a working alternative to AE. Which bypasses any smarts AE might have about format selection.
DSound and XAudio sinks enumerate and support float (the system mix format) so that behavior is not visible for them.

I added a default case for completeness but started to wonder what to do because current behavior doesn't seem right.

fritsch replied:
I now remember.

There is not much smartness needed, reason: ffmpeg decoders produce float (planar) 32 bit. So AE tries that (non planar) directly, but sink can decide to opt out. So AE will then transform the format.

For passthrough it is checked if sink supports stream type raw and int16.

That way a platform sink can overrule any potential AE smartness and logic inside AE is totally simple.

CrystalP replied:
OK. In light of that information I'll withdraw that commit as the 8 bit format doesn't make much sense either.

CrystalP committed:
[XAudio2] fix device name used to find fallback format in sink initialization

CrystalP commented:
Removed commit about data formats, added a commit for other problem spotted with the device name in InitializeInternal().
In the initial check for the AE provided format, there is some logic to adjust the device passed to CreateMasteringVoice()) but those adjustments are forgotten in later calls.

CrystalP commented:
Found some documentation about the audio categories, for Windows 8 ForegroundOnlyMedia should be more appropriate than Other, though the actual effect is not clear. Modified the first commit.

CrystalP commented:
Jenkins build this please
